crates/dbsp/src/operator/apply3.rs

```diff
@@ -0,0 +1,198 @@
+        &self,
+        other1: &Stream<C, T2>,
+        other2: &Stream<C, T3>,
+        func: F,
+    ) -> Stream<C, T4>
+    where
+        T2: Clone + 'static,
+        T3: Clone + 'static,
+        T4: Clone + 'static,
+        F: Fn(T1, &T2, &T3) -> T4 + 'static,
+    {
+        self.circuit().add_ternary_operator_with_preference(
+            Apply3Owned::new(func, Location::caller()),
+            (self, OwnershipPreference::STRONGLY_PREFER_OWNED),
+            (other1, OwnershipPreference::INDIFFERENT),
+            (other2, OwnershipPreference::INDIFFERENT),
+        )
+    }
+}
+
+/// Applies a user-provided ternary function to its inputs at each timestamp.
+pub struct Apply3<F> {
+    func: F,
+    location: &'static Location<'static>,
+}
+
+impl<F> Apply3<F> {
+    pub const fn new(func: F, location: &'static Location<'static>) -> Self
+    where
+        F: 'static,
+    {
+        Self { func, location }
+    }
+}
+
+impl<F> Operator for Apply3<F>
+where
+    F: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::Borrowed("Apply3")
+    }
+
+    fn location(&self) -> OperatorLocation {
+        Some(self.location)
+    }
+
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        // TODO: either change `F` type to `Fn` from `FnMut` or
+        // parameterize the operator with custom fixed point check.
+        unimplemented!();
+    }
+}
+
+impl<T1, T2, T3, T4, F> TernaryOperator<T1, T2, T3, T4> for Apply3<F>
+where
+    F: Fn(&T1, &T2, &T3) -> T4 + 'static,
+    T1: Clone,
+    T2: Clone,
+    T3: Clone,
+{
+    fn eval<'a>(&mut self, i1: Cow<'a, T1>, i2: Cow<'a, T2>, i3: Cow<'a, T3>) -> T4 {
+        (self.func)(&i1, &i2, &i3)
+    }
+}
+
+/// Applies a user-provided ternary function to its inputs at each timestamp,
+/// consuming the first input.
+pub struct Apply3Owned<F> {
+    func: F,
+    location: &'static Location<'static>,
+}
+
+impl<F> Apply3Owned<F> {
+    pub const fn new(func: F, location: &'static Location<'static>) -> Self
+    where
+        F: 'static,
+    {
+        Self { func, location }
+    }
+}
+
+impl<F> Operator for Apply3Owned<F>
+where
+    F: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::Borrowed("Apply3Owned")
+    }
+
+    fn location(&self) -> OperatorLocation {
+        Some(self.location)
+    }
+
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        // TODO: either change `F` type to `Fn` from `FnMut` or
+        // parameterize the operator with custom fixed point check.
+        unimplemented!();
+    }
+}
+
+impl<T1, T2, T3, T4, F> TernaryOperator<T1, T2, T3, T4> for Apply3Owned<F>
+where
+    F: Fn(T1, &T2, &T3) -> T4 + 'static,
+    T1: Clone,
+    T2: Clone,
+    T3: Clone,
+{
+    fn eval<'a>(&mut self, i1: Cow<'a, T1>, i2: Cow<'a, T2>, i3: Cow<'a, T3>) -> T4 {
+        debug_assert!(matches!(i1, Cow::Owned(_)));
+        (self.func)(i1.into_owned(), &i2, &i3)
+    }
+
+    fn input_preference(
+        &self,
```

Comment:
Doesn't need a &self?

Reply:
Should this be used in `add_ternary_operator_with_preference`

Reply:
It's part of the trait, and can in theory require `self`. 

> Should this be used in add_ternary_operator_with_preference

This is the default. `add_ternary_operator_with_preference` overrides the default.